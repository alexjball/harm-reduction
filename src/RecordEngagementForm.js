import React, { useReducer, useEffect, useState } from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import FormSelect from './FormSelect';
import { db } from './Firebase';

export default () => {
  const [state, update] = useReducer(
    (state, update) => ({ ...state, ...update }),
    {
      participantId: '',
      firstName: '',
      lastName: '',
      associatedIncident: 'None selected',
      dateEngaged: '',
      pointPerson: '',
      stateOfChange: '',
      needsIdentified: '',
      narcanEnrollment: '',
      followUpDate: '',
      firstPerson: '',
      notes: '',
      status: 'Waiting for input',
    }
  );
  const [incidents, setIncidents] = useState([]);
  const [participants, setParticipants] = useState([]);

  useEffect(
    () =>
      db.collection('incidents').onSnapshot(snapshot => {
        const incidents = [];
        snapshot.forEach(doc => {
          incidents.push({
            id: doc.id,
            ...doc.data(),
          });
        });
        setIncidents(incidents);
      }),
    []
  );

  useEffect(
    () =>
      db.collection('participants').onSnapshot(snapshot => {
        const participants = [];
        snapshot.forEach(doc => {
          participants.push({
            id: doc.id,
            ...doc.data(),
          });
        });
        setParticipants(participants);
      }),
    []
  );

  const handleChange = e => {
    const value = e.target.value;
    const id = e.target.id;
    update({ [id]: value });
  };

  const handleSubmit = async event => {
    event.preventDefault();
    event.stopPropagation();

    try {
      await db.collection('engagements').add({
        participantId: state.participantId,
        firstName: state.firstName,
        lastName: state.lastName,
        associatedIncident: state.associatedIncident,
        dateEngaged: state.dateEngaged,
        pointPerson: state.pointPerson,
        stateOfChange: state.stateOfChange,
        needsIdentified: state.needsIdentified,
        narcanEnrollment: state.narcanEnrollment,
        followUpDate: state.followUpDate,
        firstPerson: state.firstPerson,
        notes: state.notes,
      });
      update({ status: 'Submitted!' });
    } catch (e) {
      update({ status: 'Error! ' + e });
    }
  };

  const displayIncident = incident => {
    return (
      <Form.Group
        style={{
          border: '.5px',
          borderStyle: 'solid',
          borderColor: 'lightgray',
          borderRadius: '5px',
          padding: '1em',
        }}
      >
        <p>Incident Id: {incident.id}</p>
        <p>Date of Incident: {incident.dateOfRequest}</p>
        <p>Location: {incident.location}</p>
        <p>Recieved Narcan: {incident.receivedNarcan}</p>
        <p>ParticipantId: {incident.participantId}</p>
      </Form.Group>
    );
  };
  const fillParticipantInfo = e => {
    const partFiltered = participants.filter(a => a.id === e.target.value);
    update({ participantId: e.target.value });
    update({ firstName: partFiltered[0] ? partFiltered[0].firstName : '' });
    update({ lastName: partFiltered[0] ? partFiltered[0].lastName : '' });
  };

  const associatedIncidents = incidents.filter(
    a => a.participantId === state.participantId
  );

  return (
    <div>
      <div
        style={{
          width: '50%',
          marginLeft: '5%',
          marginTop: '5%',
          marginBottom: '5%',
        }}
      >
        <div
          style={{
            display: 'flex',
            alignItems: 'baseline',
            justifyContent: 'flex-start',
            marginBottom: '2em',
          }}
        >
          <h2 style={{ flex: 1 }}>Add a new engagement</h2>
          <p style={{ flex: 1, fontWeight: '400', fontSize: '1em' }}>
            <em>{state.status}</em>
          </p>
        </div>
        <Form onSubmit={handleSubmit} style={{ marginBottom: '20px' }}>
          <Row>
            <Form.Group as={Col} controlId="firstName">
              <Form.Label>First Name</Form.Label>
              <Form.Control
                value={state.firstName}
                onChange={e => update({ firstName: e.target.value })}
              />
            </Form.Group>
            <Form.Group as={Col} controlId="lastName">
              <Form.Label>Last Name</Form.Label>
              <Form.Control
                value={state.lastName}
                type="text"
                onChange={e => update({ lastName: e.target.value })}
              />
            </Form.Group>
          </Row>
          <Form.Group controlId="participantId">
            <Form.Label>Participant ID</Form.Label>
            <Form.Control
              required
              value={state.participantId}
              type="text"
              onChange={e => fillParticipantInfo(e)}
            />
          </Form.Group>
          <FormSelect
            label="Associated Incident"
            options={['None selected', ...associatedIncidents.map(a => a.id)]}
            onChange={handleChange}
          />
          <Form.Group>
            {associatedIncidents
              .filter(a => a.id === state.associatedIncident)
              .map(a => displayIncident(a))}
          </Form.Group>
          <Form.Check
            id="narcanEnrollment"
            style={{ margin: '20px 0 10px 0' }}
            label="Narcan Enrollment"
            checked={state.narcanEnrollment}
            type="checkbox"
            onChange={e => update({ narcanEnrollment: e.target.checked })}
          />
          <Form.Check
            id="firstPerson"
            style={{ margin: '10px 0 20px 0' }}
            label="First Person"
            checked={state.firstPerson}
            type="checkbox"
            onChange={e => update({ firstPerson: e.target.checked })}
          />
          <Form.Group controlId="dateEngaged">
            <Form.Label>Date Engaged</Form.Label>
            <Form.Control
              required
              value={state.dateEngaged}
              type="date"
              onChange={e => update({ dateEngaged: e.target.value })}
            />
          </Form.Group>
          <FormSelect
            label="Point Person"
            value={state.pointPerson}
            onChange={handleChange}
            options={['Not Assigned', 'CarrieAnn', 'Chris']}
          />
          <FormSelect
            label="Stage of Change"
            value={state.stageOfChange}
            onChange={handleChange}
            options={[
              'Not Decided',
              'Precontemplation',
              'Contemplation',
              'Preparation',
              'Action',
              'Maintenance',
            ]}
          />
          <Form.Group controlId="followUpDate">
            <Form.Label>Follow Up Date</Form.Label>
            <Form.Control
              required
              value={state.followUpDate}
              type="date"
              onChange={e => update({ followUpDate: e.target.value })}
            />
          </Form.Group>
          <Form.Group controlId="notes">
            <Form.Label>Notes</Form.Label>
            <Form.Control as="textarea" rows="7" />
          </Form.Group>
          <Button variant="primary" type="submit">
            Record Engagement
          </Button>{' '}
          STATUS: {state.status}
        </Form>
      </div>
    </div>
  );
};
