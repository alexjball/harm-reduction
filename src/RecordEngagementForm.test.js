import { render, getByText } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import RecordEngagementForm from './RecordEngagementForm';

jest.mock('./API');

test('Renders Incident Form', () => {
  const engagementForm = render(<RecordEngagementForm />);
  expect(engagementForm).toMatchSnapshot();
});

test.each([
  ['First Name', 'Name', 'Name'],
  ['Last Name', 'Name', 'Name'],
])('%s takes input', (a, b, expected) => {
  const { getByLabelText } = render(<RecordEngagementForm />);
  userEvent.type(getByLabelText(a), b);
  expect(getByLabelText(a)).toHaveValue(expected);
});

//TODO test participant ID, associated incident

test.each([['Narcan Enrollment'], ['First Person']])('%s can be checked', a => {
  const { getByLabelText } = render(<RecordEngagementForm />);

  userEvent.click(getByLabelText(a));
  expect(getByLabelText(a)).toBeChecked();
});

test('Sets stage of change', () => {
  const { getByLabelText } = render(<RecordEngagementForm />);

  userEvent.selectOptions(
    getByLabelText('Stage of Change'),
    'Precontemplation'
  );

  expect(getByLabelText('Stage of Change')).toHaveValue('Precontemplation');
});

test('Date Engaged accepts date', async () => {
  const { getByLabelText } = render(<RecordEngagementForm />);
  const dateInput = getByLabelText('Date Engaged');
  const date = '2020-12-12';

  await userEvent.type(dateInput, date);

  expect(dateInput).toHaveValue(date);
});

test('Can select point person', () => {
  const { getByLabelText } = render(<RecordEngagementForm />);
  userEvent.selectOptions(getByLabelText('Point Person'), 'CarrieAnn');
  expect(getByLabelText('Point Person')).toHaveValue('CarrieAnn');
});
