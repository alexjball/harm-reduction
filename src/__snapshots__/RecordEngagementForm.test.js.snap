// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Incident Form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          style="width: 50%; margin-left: 5%; margin-top: 5%; margin-bottom: 5%;"
        >
          <div
            style="display: flex; align-items: baseline; justify-content: flex-start; margin-bottom: 2em;"
          >
            <h2
              style="flex: 1;"
            >
              Add a new engagement
            </h2>
            <p
              style="flex: 1; font-weight: 400; font-size: 1em;"
            >
              <em>
                Waiting for input
              </em>
            </p>
          </div>
          <form
            class=""
            style="margin-bottom: 20px;"
          >
            <div
              class="row"
            >
              <div
                class="form-group col"
              >
                <label
                  class="form-label"
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  class="form-control"
                  id="firstName"
                  value=""
                />
              </div>
              <div
                class="form-group col"
              >
                <label
                  class="form-label"
                  for="lastName"
                >
                  Last Name
                </label>
                <input
                  class="form-control"
                  id="lastName"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="participantId"
              >
                Participant ID
              </label>
              <input
                class="form-control"
                id="participantId"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="associatedIncident"
              >
                Associated Incident
              </label>
              <select
                class="form-control"
                id="associatedIncident"
                type="text"
              >
                <option>
                  None selected
                </option>
              </select>
            </div>
            <div
              class="form-check"
              style="margin: 20px 0px 10px 0px;"
            >
              <input
                class="form-check-input"
                id="narcanEnrollment"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="narcanEnrollment"
                title=""
              >
                Narcan Enrollment
              </label>
            </div>
            <div
              class="form-check"
              style="margin: 10px 0px 20px 0px;"
            >
              <input
                class="form-check-input"
                id="firstPerson"
                type="checkbox"
              />
              <label
                class="form-check-label"
                for="firstPerson"
                title=""
              >
                First Person
              </label>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="dateEngaged"
              >
                Date Engaged
              </label>
              <input
                class="form-control"
                id="dateEngaged"
                required=""
                type="date"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="pointPerson"
              >
                Point Person
              </label>
              <select
                class="form-control"
                id="pointPerson"
                type="text"
              >
                <option>
                  Not Assigned
                </option>
                <option>
                  CarrieAnn
                </option>
                <option>
                  Chris
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="stageOfChange"
              >
                Stage of Change
              </label>
              <select
                class="form-control"
                id="stageOfChange"
                type="text"
              >
                <option>
                  Not Decided
                </option>
                <option>
                  Precontemplation
                </option>
                <option>
                  Contemplation
                </option>
                <option>
                  Preparation
                </option>
                <option>
                  Action
                </option>
                <option>
                  Maintenance
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="followUpDate"
              >
                Follow Up Date
              </label>
              <input
                class="form-control"
                id="followUpDate"
                required=""
                type="date"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="notes"
              >
                Notes
              </label>
              <textarea
                class="form-control"
                id="notes"
                rows="7"
              />
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Record Engagement
            </button>
             
            STATUS: 
            Waiting for input
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        style="width: 50%; margin-left: 5%; margin-top: 5%; margin-bottom: 5%;"
      >
        <div
          style="display: flex; align-items: baseline; justify-content: flex-start; margin-bottom: 2em;"
        >
          <h2
            style="flex: 1;"
          >
            Add a new engagement
          </h2>
          <p
            style="flex: 1; font-weight: 400; font-size: 1em;"
          >
            <em>
              Waiting for input
            </em>
          </p>
        </div>
        <form
          class=""
          style="margin-bottom: 20px;"
        >
          <div
            class="row"
          >
            <div
              class="form-group col"
            >
              <label
                class="form-label"
                for="firstName"
              >
                First Name
              </label>
              <input
                class="form-control"
                id="firstName"
                value=""
              />
            </div>
            <div
              class="form-group col"
            >
              <label
                class="form-label"
                for="lastName"
              >
                Last Name
              </label>
              <input
                class="form-control"
                id="lastName"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="participantId"
            >
              Participant ID
            </label>
            <input
              class="form-control"
              id="participantId"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="associatedIncident"
            >
              Associated Incident
            </label>
            <select
              class="form-control"
              id="associatedIncident"
              type="text"
            >
              <option>
                None selected
              </option>
            </select>
          </div>
          <div
            class="form-check"
            style="margin: 20px 0px 10px 0px;"
          >
            <input
              class="form-check-input"
              id="narcanEnrollment"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="narcanEnrollment"
              title=""
            >
              Narcan Enrollment
            </label>
          </div>
          <div
            class="form-check"
            style="margin: 10px 0px 20px 0px;"
          >
            <input
              class="form-check-input"
              id="firstPerson"
              type="checkbox"
            />
            <label
              class="form-check-label"
              for="firstPerson"
              title=""
            >
              First Person
            </label>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="dateEngaged"
            >
              Date Engaged
            </label>
            <input
              class="form-control"
              id="dateEngaged"
              required=""
              type="date"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="pointPerson"
            >
              Point Person
            </label>
            <select
              class="form-control"
              id="pointPerson"
              type="text"
            >
              <option>
                Not Assigned
              </option>
              <option>
                CarrieAnn
              </option>
              <option>
                Chris
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="stageOfChange"
            >
              Stage of Change
            </label>
            <select
              class="form-control"
              id="stageOfChange"
              type="text"
            >
              <option>
                Not Decided
              </option>
              <option>
                Precontemplation
              </option>
              <option>
                Contemplation
              </option>
              <option>
                Preparation
              </option>
              <option>
                Action
              </option>
              <option>
                Maintenance
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="followUpDate"
            >
              Follow Up Date
            </label>
            <input
              class="form-control"
              id="followUpDate"
              required=""
              type="date"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="notes"
            >
              Notes
            </label>
            <textarea
              class="form-control"
              id="notes"
              rows="7"
            />
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Record Engagement
          </button>
           
          STATUS: 
          Waiting for input
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
